File: shared/states/go/states.go
// Package states is the canonical Go binding of the video FSM owned by
// story-01-06. The 12 video states, the 7 processing-stage names, and
// the transition graph are mirrored from shared/states/states.json
// and pinned by tests so they cannot drift from the Python binding in
// pipeline/src/maktaba_pipeline/domain/states.py or from the migration
// CHECK constraint in shared/db/migrations/0004_video_states_and_stages.sql.
//
// AdvanceAfterStage is the SOLE function callers should use to mutate
// videos.state. It validates the (from, trigger, outcome) triple against
// the runtime transition table and returns IllegalStateTransition for
// anything outside the allowed set. See plan-01-06 §6.4 for why the DB
// only enforces set membership and not transition validity.
package states

// State is the canonical video-state enum. The string values are the
// values that land in videos.state and that the slot 0004 CHECK
// constraint enumerates.
type State string

// The 12 canonical video states. Iteration order in AllStates matches
// the order in shared/states/states.json so tests can compare directly.
const (
	StateDiscovered     State = "discovered"
	StateProbed         State = "probed"
	StateAudioExtracted State = "audio_extracted"
	StateTranscribed    State = "transcribed"
	StateIndexed        State = "indexed"
	StateThumbnailed    State = "thumbnailed"
	StateReady          State = "ready"
	StateReadyNoAudio   State = "ready_no_audio"
	StateMissing        State = "missing"
	StateSuperseded     State = "superseded"
	StateCorrupted      State = "corrupted"
	StateFailed         State = "failed"
)

// AllStates is the full set in manifest order.
var AllStates = []State{
	StateDiscovered,
	StateProbed,
	StateAudioExtracted,
	StateTranscribed,
	StateIndexed,
	StateThumbnailed,
	StateReady,
	StateReadyNoAudio,
	StateMissing,
	StateSuperseded,
	StateCorrupted,
	StateFailed,
}

// Class returns the FSM class of a state — one of "open",
// "terminal-good", "terminal-soft", "terminal-bad", or "sink".
func (s State) Class() string { return classOf[s] }

// IsTerminalDrop reports whether a state should silently drop incoming
// stage finishes. SUPERSEDED, CORRUPTED, and FAILED rows are
// terminal-soft / terminal-bad: a worker that finishes after the row
// hit one of these states is racing with a side-channel transition;
// AdvanceAfterStage logs and no-ops instead of erroring.
func (s State) IsTerminalDrop() bool {
	switch s {
	case StateSuperseded, StateCorrupted, StateFailed:
		return true
	default:
		return false
	}
}

var classOf = map[State]string{
	StateDiscovered:     "open",
	StateProbed:         "open",
	StateAudioExtracted: "open",
	StateTranscribed:    "open",
	StateIndexed:        "open",
	StateThumbnailed:    "open",
	StateReady:          "terminal-good",
	StateReadyNoAudio:   "terminal-good",
	StateMissing:        "sink",
	StateSuperseded:     "terminal-soft",
	StateCorrupted:      "terminal-bad",
	StateFailed:         "terminal-bad",
}

// Stage is the canonical processing_jobs.stage enum.
type Stage string

const (
	StageScan         Stage = "scan"
	StageProbe        Stage = "probe"
	StageExtract      Stage = "extract"
	StageTranscribe   Stage = "transcribe"
	StageSubtitleGen  Stage = "subtitle_gen"
	StageIndex        Stage = "index"
	StageThumbnail    Stage = "thumbnail"
)

// AllStages is the full set in manifest order. The slot 0002 + slot
// 0004 CHECK constraints enumerate the same seven values.
var AllStages = []Stage{
	StageScan,
	StageProbe,
	StageExtract,
	StageTranscribe,
	StageSubtitleGen,
	StageIndex,
	StageThumbnail,
}

// Trigger is a superset of Stage that includes the three side-channel
// triggers (filesystem, library, integrity) carrying the "any-state"
// transitions to MISSING / SUPERSEDED / CORRUPTED. Triggers are an
// in-process concept; only the seven Stage values appear in the DB.
type Trigger string

const (
	TriggerScan         Trigger = "scan"
	TriggerProbe        Trigger = "probe"
	TriggerExtract      Trigger = "extract"
	TriggerTranscribe   Trigger = "transcribe"
	TriggerSubtitleGen  Trigger = "subtitle_gen"
	TriggerIndex        Trigger = "index"
	TriggerThumbnail    Trigger = "thumbnail"
	TriggerFilesystem   Trigger = "filesystem"
	TriggerLibrary      Trigger = "library"
	TriggerIntegrity    Trigger = "integrity"
)

// AllTriggers is the full ten-element superset.
var AllTriggers = []Trigger{
	TriggerScan,
	TriggerProbe,
	TriggerExtract,
	TriggerTranscribe,
	TriggerSubtitleGen,
	TriggerIndex,
	TriggerThumbnail,
	TriggerFilesystem,
	TriggerLibrary,
	TriggerIntegrity,
}

// IsStage reports whether a trigger corresponds to one of the seven
// canonical pipeline stages (i.e. not a side-channel trigger).
func (t Trigger) IsStage() bool {
	switch t {
	case TriggerScan, TriggerProbe, TriggerExtract, TriggerTranscribe,
		TriggerSubtitleGen, TriggerIndex, TriggerThumbnail:
		return true
	default:
		return false
	}
}

// Outcome is the result-token a caller passes alongside a trigger.
// Open at the type level; the runtime transition table enumerates the
// outcomes the FSM understands. Anything else is an
// IllegalStateTransition even from an otherwise-compatible source.
type Outcome string

// Recognized outcomes. Listed for documentation; new outcomes only
// matter when added to the transition table below.
const (
	OutcomeOK            Outcome = "ok"
	OutcomeNoAudio       Outcome = "no_audio"
	OutcomePartial       Outcome = "partial"
	OutcomeExhausted     Outcome = "exhausted"
	OutcomeRediscovered  Outcome = "rediscovered"
	OutcomeDeleted       Outcome = "deleted"
	OutcomeReplaced      Outcome = "replaced"
	OutcomeFail          Outcome = "fail"
	OutcomeAllGatesOK    Outcome = "all_gates_ok"
)

// transitionKey is the lookup key into the explicit transition table.
type transitionKey struct {
	From    State
	Trigger Trigger
	Outcome Outcome
}

// explicitTransitions are the 11 hand-named edges from plan-01-06 §4.
// Broadcast edges are evaluated by Lookup against state classes and
// trigger groups so we don't have to enumerate ~70 expanded rows here.
var explicitTransitions = map[transitionKey]State{
	{StateDiscovered, TriggerProbe, OutcomeOK}:                 StateProbed,
	{StateProbed, TriggerExtract, OutcomeOK}:                   StateAudioExtracted,
	{StateProbed, TriggerProbe, OutcomeNoAudio}:                StateReadyNoAudio,
	{StateAudioExtracted, TriggerTranscribe, OutcomeOK}:        StateTranscribed,
	{StateTranscribed, TriggerSubtitleGen, OutcomePartial}:     StateTranscribed,
	{StateTranscribed, TriggerIndex, OutcomePartial}:           StateTranscribed,
	{StateTranscribed, TriggerSubtitleGen, OutcomeOK}:          StateIndexed,
	{StateTranscribed, TriggerIndex, OutcomeOK}:                StateIndexed,
	{StateIndexed, TriggerThumbnail, OutcomeOK}:                StateThumbnailed,
	{StateThumbnailed, TriggerScan, OutcomeAllGatesOK}:         StateReady,
	{StateMissing, TriggerScan, OutcomeRediscovered}:           StateDiscovered,
}

// Lookup returns the target state for (from, trigger, outcome) and
// reports whether a matching edge exists. It first consults the
// explicit table, then the four broadcast rules. A miss on both is
// what AdvanceAfterStage turns into IllegalStateTransition.
func Lookup(from State, trigger Trigger, outcome Outcome) (State, bool) {
	if to, ok := explicitTransitions[transitionKey{from, trigger, outcome}]; ok {
		return to, true
	}
	return broadcastLookup(from, trigger, outcome)
}

// broadcastLookup expands the four "any-source" rows from the manifest:
//
//   - filesystem/deleted  → MISSING     (open ∪ terminal-good ∪ sink)
//   - <stage>/exhausted   → FAILED      (open ∪ terminal-good)
//   - integrity/fail      → CORRUPTED   (open ∪ terminal-good)
//   - library/replaced    → SUPERSEDED  (open ∪ terminal-good ∪ terminal-soft)
//
// The stage-wildcard exhausted edge matches any of the seven canonical
// stage triggers, never a side-channel trigger.
func broadcastLookup(from State, trigger Trigger, outcome Outcome) (State, bool) {
	cls := from.Class()

	switch {
	case trigger == TriggerFilesystem && outcome == OutcomeDeleted:
		if cls == "open" || cls == "terminal-good" || cls == "sink" {
			return StateMissing, true
		}

	case trigger.IsStage() && outcome == OutcomeExhausted:
		if cls == "open" || cls == "terminal-good" {
			return StateFailed, true
		}

	case trigger == TriggerIntegrity && outcome == OutcomeFail:
		if cls == "open" || cls == "terminal-good" {
			return StateCorrupted, true
		}

	case trigger == TriggerLibrary && outcome == OutcomeReplaced:
		if cls == "open" || cls == "terminal-good" || cls == "terminal-soft" {
			return StateSuperseded, true
		}
	}

	return "", false
}

File: pipeline/src/maktaba_pipeline/domain/states.py
"""Canonical video FSM — the Python binding of Story 1.6.

Source of truth: ``shared/states/states.json``. The Go binding under
``shared/states/go`` mirrors this module one-for-one; tests in both
trees pin the parity.

The 12 :class:`State` values are exactly the strings the slot 0004
``videos_state_valid`` CHECK constraint accepts. The 7 :class:`Stage`
values are exactly the strings the slot 0002/0004 stage CHECKs accept.
The :class:`Trigger` enum is a 10-element superset that adds three
side-channel triggers (``filesystem``, ``library``, ``integrity``) for
the broadcast transitions.

Use :func:`lookup` to resolve a ``(from, trigger, outcome)`` triple to
its target state. ``None`` means the triple is not in the transition
graph; callers should raise :class:`IllegalStateTransition` on that
result. The orchestrator's :func:`maktaba_pipeline.orchestrator.advance.
advance_after_stage` is the only function that should ever issue an
``UPDATE videos SET state = …`` statement.
"""

from __future__ import annotations

import enum
from dataclasses import dataclass

__all__ = [
    "BROADCAST_TRANSITIONS",
    "EXPLICIT_TRANSITIONS",
    "IllegalStateTransition",
    "Outcome",
    "STATE_CLASS",
    "Stage",
    "State",
    "TERMINAL_DROP_STATES",
    "Trigger",
    "allowed_targets",
    "allowed_transitions",
    "is_stage_trigger",
    "lookup",
]


class State(enum.StrEnum):
    """The 12 canonical video states.

    Mirrors the ``videos_state_valid`` CHECK constraint added by slot
    0004. The class hierarchy (open / terminal-good / terminal-soft /
    terminal-bad / sink) drives the broadcast transition rules in
    :data:`BROADCAST_TRANSITIONS`.
    """

    DISCOVERED = "discovered"
    PROBED = "probed"
    AUDIO_EXTRACTED = "audio_extracted"
    TRANSCRIBED = "transcribed"
    INDEXED = "indexed"
    THUMBNAILED = "thumbnailed"
    READY = "ready"
    READY_NO_AUDIO = "ready_no_audio"
    MISSING = "missing"
    SUPERSEDED = "superseded"
    CORRUPTED = "corrupted"
    FAILED = "failed"


class Stage(enum.StrEnum):
    """The 7 canonical pipeline stage names.

    Mirrors ``processing_jobs_stage_chk`` (slot 0002) and
    ``processing_jobs_stage_valid`` (slot 0004); both enforce the same
    set. Adding a new stage requires (a) a new value here, (b) a
    follow-up migration that ALTERs both CHECKs, and (c) extending the
    transition table below.
    """

    SCAN = "scan"
    PROBE = "probe"
    EXTRACT = "extract"
    TRANSCRIBE = "transcribe"
    SUBTITLE_GEN = "subtitle_gen"
    INDEX = "index"
    THUMBNAIL = "thumbnail"


class Trigger(enum.StrEnum):
    """Superset of :class:`Stage` plus three side-channel triggers.

    Side-channel triggers (``filesystem``, ``library``, ``integrity``)
    carry the broadcast transitions to ``MISSING``, ``SUPERSEDED``, and
    ``CORRUPTED`` respectively. They never appear in
    ``processing_jobs.stage`` — that column rejects anything outside
    the 7 :class:`Stage` values.
    """

    SCAN = "scan"
    PROBE = "probe"
    EXTRACT = "extract"
    TRANSCRIBE = "transcribe"
    SUBTITLE_GEN = "subtitle_gen"
    INDEX = "index"
    THUMBNAIL = "thumbnail"
    FILESYSTEM = "filesystem"
    LIBRARY = "library"
    INTEGRITY = "integrity"


# Recognized outcomes from plan-01-06 §3.4. Open vocabulary at the type
# level; closed at the runtime-table level. Anything not listed in
# EXPLICIT_TRANSITIONS or matched by a broadcast row raises
# IllegalStateTransition.
class Outcome(enum.StrEnum):
    OK = "ok"
    NO_AUDIO = "no_audio"
    PARTIAL = "partial"
    EXHAUSTED = "exhausted"
    REDISCOVERED = "rediscovered"
    DELETED = "deleted"
    REPLACED = "replaced"
    FAIL = "fail"
    ALL_GATES_OK = "all_gates_ok"


# Class assignment matches shared/states/states.json. The class
# determines which broadcast rows can fire from a given state.
STATE_CLASS: dict[State, str] = {
    State.DISCOVERED: "open",
    State.PROBED: "open",
    State.AUDIO_EXTRACTED: "open",
    State.TRANSCRIBED: "open",
    State.INDEXED: "open",
    State.THUMBNAILED: "open",
    State.READY: "terminal-good",
    State.READY_NO_AUDIO: "terminal-good",
    State.MISSING: "sink",
    State.SUPERSEDED: "terminal-soft",
    State.CORRUPTED: "terminal-bad",
    State.FAILED: "terminal-bad",
}


# Terminal-drop states: a stage that finishes after the row hit one of
# these is racing with a side-channel transition. advance_after_stage
# logs `late_stage_finish` and no-ops instead of erroring (story edge
# case 1).
TERMINAL_DROP_STATES: frozenset[State] = frozenset(
    {State.SUPERSEDED, State.CORRUPTED, State.FAILED}
)


def is_stage_trigger(t: Trigger) -> bool:
    """True iff ``t`` is one of the 7 canonical pipeline stages."""
    return t.value in {s.value for s in Stage}


@dataclass(frozen=True, slots=True)
class _Key:
    """Triple key into the explicit transition table."""

    frm: State
    trigger: Trigger
    outcome: str


# The 11 hand-named edges from plan-01-06 §4. Broadcast rows are
# evaluated separately by :func:`lookup` so we don't enumerate ~70
# expanded rows here.
EXPLICIT_TRANSITIONS: dict[_Key, State] = {
    _Key(State.DISCOVERED, Trigger.PROBE, Outcome.OK.value): State.PROBED,
    _Key(State.PROBED, Trigger.EXTRACT, Outcome.OK.value): State.AUDIO_EXTRACTED,
    _Key(State.PROBED, Trigger.PROBE, Outcome.NO_AUDIO.value): State.READY_NO_AUDIO,
    _Key(State.AUDIO_EXTRACTED, Trigger.TRANSCRIBE, Outcome.OK.value): State.TRANSCRIBED,
    _Key(State.TRANSCRIBED, Trigger.SUBTITLE_GEN, Outcome.PARTIAL.value): State.TRANSCRIBED,
    _Key(State.TRANSCRIBED, Trigger.INDEX, Outcome.PARTIAL.value): State.TRANSCRIBED,
    _Key(State.TRANSCRIBED, Trigger.SUBTITLE_GEN, Outcome.OK.value): State.INDEXED,
    _Key(State.TRANSCRIBED, Trigger.INDEX, Outcome.OK.value): State.INDEXED,
    _Key(State.INDEXED, Trigger.THUMBNAIL, Outcome.OK.value): State.THUMBNAILED,
    _Key(State.THUMBNAILED, Trigger.SCAN, Outcome.ALL_GATES_OK.value): State.READY,
    _Key(State.MISSING, Trigger.SCAN, Outcome.REDISCOVERED.value): State.DISCOVERED,
}


# Broadcast rules — each row matches a (trigger, outcome) pair from any
# source state whose class is in ``source_classes``. The Stage wildcard
# only matches the 7 canonical stage triggers (never side-channel ones).
@dataclass(frozen=True, slots=True)
class _BroadcastRow:
    source_classes: frozenset[str]
    trigger: Trigger | None  # None means "any stage trigger"
    outcome: str
    to: State


BROADCAST_TRANSITIONS: tuple[_BroadcastRow, ...] = (
    _BroadcastRow(
        source_classes=frozenset({"open", "terminal-good", "sink"}),
        trigger=Trigger.FILESYSTEM,
        outcome=Outcome.DELETED.value,
        to=State.MISSING,
    ),
    _BroadcastRow(
        source_classes=frozenset({"open", "terminal-good"}),
        trigger=None,  # any stage
        outcome=Outcome.EXHAUSTED.value,
        to=State.FAILED,
    ),
    _BroadcastRow(
        source_classes=frozenset({"open", "terminal-good"}),
        trigger=Trigger.INTEGRITY,
        outcome=Outcome.FAIL.value,
        to=State.CORRUPTED,
    ),
    _BroadcastRow(
        source_classes=frozenset({"open", "terminal-good", "terminal-soft"}),
        trigger=Trigger.LIBRARY,
        outcome=Outcome.REPLACED.value,
        to=State.SUPERSEDED,
    ),
)


def lookup(frm: State, trigger: Trigger, outcome: str) -> State | None:
    """Resolve ``(frm, trigger, outcome)`` to a target state.

    Returns ``None`` if the triple is not in the transition graph. The
    explicit table is checked first; on a miss the broadcast rules are
    evaluated. Outcome is taken as a string so callers can pass either
    an :class:`Outcome` value or a literal string (the JSON manifest
    uses strings).
    """
    explicit = EXPLICIT_TRANSITIONS.get(_Key(frm, trigger, outcome))
    if explicit is not None:
        return explicit

    cls = STATE_CLASS[frm]
    for row in BROADCAST_TRANSITIONS:
        if cls not in row.source_classes:
            continue
        if row.outcome != outcome:
            continue
        if row.trigger is None:
            # Stage-wildcard row — only stage triggers may match.
            if is_stage_trigger(trigger):
                return row.to
            continue
        if row.trigger.value == trigger.value:
            return row.to
    return None


# View dict[State, set[State]] — the simpler shape the story names.
# Built lazily and cached so the module's import-time cost stays low.
allowed_targets: dict[State, frozenset[State]]


def _build_allowed_targets() -> dict[State, frozenset[State]]:
    out: dict[State, set[State]] = {s: set() for s in State}
    for key, target in EXPLICIT_TRANSITIONS.items():
        out[key.frm].add(target)
    for src in State:
        cls = STATE_CLASS[src]
        for row in BROADCAST_TRANSITIONS:
            if cls in row.source_classes:
                out[src].add(row.to)
    return {s: frozenset(t) for s, t in out.items()}


allowed_targets = _build_allowed_targets()


# `allowed_transitions` is the richer triple-keyed table the
# orchestrator uses; the simpler `allowed_targets` view satisfies the
# story's `dict[State, set[State]]` shape.
allowed_transitions: dict[_Key, State] = EXPLICIT_TRANSITIONS


class IllegalStateTransition(Exception):
    """Raised when a (from, trigger, outcome) triple is not in the FSM.

    Carries the triple so observability can log it; callers that want
    to dispatch on the source state can read ``self.from_``.
    """

    def __init__(
        self,
        video_id: object,
        from_: State,
        trigger: Trigger,
        outcome: str,
    ) -> None:
        self.video_id = video_id
        self.from_ = from_
        self.trigger = trigger
        self.outcome = outcome
        super().__init__(
            f"illegal state transition for video {video_id}: "
            f"from={from_.value} trigger={trigger.value} outcome={outcome}"
        )
